File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        Map<Integer, Integer> countMap = new HashMap<>();

        // 1. 각 스테이지에 머무른 사람 수 카운트
        for (int stage : stages) {
            if (stage <= N) {
                countMap.put(stage, countMap.getOrDefault(stage, 0) + 1);
            }
        }

        // 2. 실패율 계산 및 저장
        List<StageFail> failList = new ArrayList<>();
        int totalPlayers = stages.length;

        for (int i = 1; i <= N; i++) {
            int current = countMap.getOrDefault(i, 0);
            double failRate = totalPlayers == 0 ? 0 : (double) current / totalPlayers;
            failList.add(new StageFail(i, failRate));
            totalPlayers -= current; // 다음 스테이지로 넘어감
        }

        // 3. 실패율 기준 내림차순 정렬 (같으면 스테이지 번호 오름차순)
        Collections.sort(failList, (a, b) -> {
            if (b.fail != a.fail) {
                return Double.compare(b.fail, a.fail);
            } else {
                return Integer.compare(a.stage, b.stage);
            }
        });

        // 4. 정답 배열에 스테이지 번호만 저장
        for (int i = 0; i < N; i++) {
            answer[i] = failList.get(i).stage;
        }

        return answer;
    }

    // 실패율과 스테이지 번호를 저장할 클래스
    static class StageFail {
        int stage;
        double fail;

        StageFail(int stage, double fail) {
            this.stage = stage;
            this.fail = fail;
        }
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    var answer = [];
    var user = stages.length;
    var stageArray = new Array(N).fill(0);
    var failure = [];

    // 각 스테이지 실패자 수 계산
    for (let i = 0; i < stages.length; i++) {
        if (stages[i] > N) continue;
        stageArray[stages[i] - 1]++;
    }

    // 실패율 계산
    for (let j = 0; j < stageArray.length; j++) {
        let failCount = stageArray[j];
        let rate = user === 0 ? 0 : failCount / user;
        failure.push({ stage: j + 1, rate: rate });
        user -= failCount;
    }

    // 실패율 기준 내림차순 정렬 (동률이면 stage 번호 오름차순)
    failure.sort((a, b) => {
        if (b.rate !== a.rate) return b.rate - a.rate;
        return a.stage - b.stage;
    });

    // 정렬된 stage 번호만 추출
    answer = failure.map(obj => obj.stage);

    return answer;
}
